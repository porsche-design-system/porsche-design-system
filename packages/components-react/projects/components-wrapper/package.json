{
  "name": "@porsche-design-system/components-react",
  "version": "1.5.4",
  "scope": "@porsche-design-system",
  "author": "Dr. Ing. h.c. F. Porsche AG",
  "contributors": [
    "Stephan Schröter <stephan.schroeter@mhp.com>",
    "Marcel Bertram <marcel.bertram@porsche.de>",
    "Anna Krohmer <anna.krohmer@udg.de>",
    "Oliver Hook <oliver.hook@mhp.com>",
    "Olivia Pala <olivia.pala@udg.de>",
    "Christoph Stickel <net.christoph.stickel@mhp.com>",
    "Matthias Hägele <matthias.haegele@mhp.com>",
    "Melanie Kolar <melanie.kolar@udg.de>",
    "Bernd Hacker <bernd.hacker@mhp.com"
  ],
  "license": "SEE LICENSE IN https://designsystem.porsche.com",
  "description": "Porsche Design System Web Components React",
  "main": "./dist/cjs/index.js",
  "module": "./dist/esm/index.js",
  "typings": "./dist/types/index.d.ts",
  "files": [
    "./dist",
    "./mocks"
  ],
  "scripts": {
    "build": "yarn build:components && yarn build:mocks",
    "build:components": "yarn build:components:module && yarn build:components:cjs",
    "build:components:module": "tsc --build './tsconfig.esm.json'",
    "build:components:cjs": "tsc --build './tsconfig.cjs.json'",
    "build:mocks": "tsc --build './tsconfig.mocks.json'",
    "clean": "rm -rf ./node_modules && rm -rf ./yarn-error.log && rm -rf './dist' && rm -rf './mocks'"
  },
  "dependencies": {
    "@porsche-design-system/components-js": "1.5.4"
  },
  "peerDependencies": {
    "react": ">=16.9.0 <17.0.0",
    "react-dom": ">=16.9.0 <17.0.0"
  },
  "publishConfig": {
    "@porsche-design-system:registry": "https://porscheui.jfrog.io/porscheui/api/npm/npm-local/"
  }
}
