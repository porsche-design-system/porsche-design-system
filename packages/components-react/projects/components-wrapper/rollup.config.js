import typescript from '@rollup/plugin-typescript';
import copy from 'rollup-plugin-copy';
import generatePackageJson from 'rollup-plugin-generate-package-json';

const projectDir = 'projects/components-wrapper';
const outputDir = 'dist/components-wrapper';
const input = `${projectDir}/src/public-api.ts`;

const typescriptOpts = {
  tsconfig: `${projectDir}/tsconfig.json`,
};

const external = [
  '@porsche-design-system/components-js',
  '@porsche-design-system/components-js/partials',
  '@porsche-design-system/components-js/utilities/jss',
  'react',
  'react/jsx-runtime',
];

const packageJsonConfig = {
  baseContents: {
    main: 'index.js',
    module: 'esm/index.js',
    sideEffects: false,
  },
};

export default [
  {
    input,
    external,
    output: {
      dir: outputDir,
      format: 'cjs',
    },
    plugins: [
      typescript({
        ...typescriptOpts,
        declaration: true,
        declarationDir: outputDir,
        rootDir: 'src',
      }),
    ],
  },
  {
    input,
    external,
    output: {
      dir: `${outputDir}/esm`,
      format: 'esm',
    },
    plugins: [typescript(typescriptOpts)],
  },
  {
    input: `${projectDir}/src/partials.ts`,
    external,
    output: [
      {
        file: `${outputDir}/partials/index.js`,
        format: 'cjs',
        plugins: [generatePackageJson(packageJsonConfig)],
      },
      {
        file: `${outputDir}/partials/esm/index.js`,
        format: 'esm',
      },
    ],
    plugins: [typescript(typescriptOpts)],
  },
  {
    input: `${projectDir}/src/utilities/jss.ts`,
    external,
    output: [
      {
        file: `${outputDir}/utilities/jss/index.js`,
        format: 'cjs',
        plugins: [generatePackageJson(packageJsonConfig)],
      },
      {
        file: `${outputDir}/utilities/jss/esm/index.js`,
        format: 'esm',
      },
    ],
    plugins: [typescript(typescriptOpts)],
  },
  {
    input: `${projectDir}/src/testing.ts`,
    external: ['@testing-library/dom'],
    output: {
      dir: outputDir,
      format: 'cjs',
    },
    plugins: [
      copy({
        targets: [{ src: `${projectDir}/src/utilities/scss.scss`, dest: `${outputDir}/utilities` }],
      }),
      typescript(typescriptOpts),
    ],
  },
];
