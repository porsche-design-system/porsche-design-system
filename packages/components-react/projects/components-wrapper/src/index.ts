export { componentsReady } from '@porsche-design-system/components-js';
export * from './lib/components';
export * from './lib/types';
export { PorscheDesignSystemProvider } from './provider';
export { skipCheckForPorscheDesignSystemProviderDuringTests, useToastManager } from './hooks';
export { skipPorscheDesignSystemCDNRequestsDuringTests } from './utils';
