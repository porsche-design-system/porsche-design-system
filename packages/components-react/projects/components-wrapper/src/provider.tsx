import type { CSSProperties, PropsWithChildren } from 'react';
import { createContext, useContext, useEffect } from 'react';
import { load } from '@porsche-design-system/components-js';

export { componentsReady } from '@porsche-design-system/components-js';

type Props = { prefix?: string };

const PorscheDesignSystemContext = createContext({ prefix: '' });

export const PorscheDesignSystemProvider = ({ prefix = '', ...props }: PropsWithChildren<Props>): JSX.Element => {
  useEffect(() => {
    load({ prefix });
  }, []);

  return <PorscheDesignSystemContext.Provider value={{ prefix }} {...props} />;
};

export const usePrefix = (tagName: string): string => {
  const { prefix } = useContext(PorscheDesignSystemContext);

  if (prefix === undefined) {
    throw new Error('It appears the <PorscheDesignSystemProvider /> is missing. Make sure to wrap your App in it.');
  }
  return prefix ? `${prefix}-${tagName}` : tagName;
};

export const useEventCallback = (element: HTMLElement, eventName: string, eventHandler: (e: Event) => void): void => {
  useEffect(() => {
    if (element && eventHandler) {
      element.addEventListener(eventName, eventHandler);
      return () => element.removeEventListener(eventName, eventHandler);
    }
  }, [eventHandler]);
};

export type BaseType = {
  className?: string;
  style?: CSSProperties;
};
