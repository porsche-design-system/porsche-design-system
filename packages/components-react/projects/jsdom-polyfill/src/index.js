global.ResizeObserver = require('@juggle/resize-observer').ResizeObserver;
// workaround for WebComponentManager not working in JSDOM
window.PDS_SKIP_FETCH = true;

require('./lib/loader.cjs')
  .defineCustomElements()
  .then(() => {
    window.PDS_SKIP_FETCH = false;
  });
