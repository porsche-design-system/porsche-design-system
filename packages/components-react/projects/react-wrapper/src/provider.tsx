import type { PropsWithChildren, JSX } from 'react';
import { createContext, useEffect } from 'react';
import { load } from '@porsche-design-system/components-js';

// to warn users about missing PorscheDesignSystemProvider, we set the default values as undefined
export const PorscheDesignSystemContext = createContext<{ prefix?: string }>({
  prefix: undefined,
});

type Props = { prefix?: string };

export const PorscheDesignSystemProvider = ({ prefix = '', ...props }: PropsWithChildren<Props>): JSX.Element => {
  useEffect(() => {
    load({ prefix });
  }, []); // No runtime prefix change is supported

  return <PorscheDesignSystemContext.Provider value={{ prefix }} {...props} />;
};
