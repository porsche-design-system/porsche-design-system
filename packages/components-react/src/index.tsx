import 'react-app-polyfill/ie11';
import 'react-app-polyfill/stable';
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import '@porsche-design-system/utils/dist/visual-regression-test.css';
import { App } from './App';

ReactDOM.render(
  <Router>
    <App />
  </Router>,
  document.getElementById('root')
);
