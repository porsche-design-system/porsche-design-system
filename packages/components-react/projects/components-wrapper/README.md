# Porsche Design System - Components React

Porsche Design System is a component library designed to help developers create the best experience for software or
services distributed by Dr. Ing. h.c. F. Porsche AG. Visit the [Porsche Design System](https://designsystem.porsche.com)
to learn more.

## Using the Porsche Design System

### Installation

Run the following command using [npm](https://www.npmjs.com):

```bash
npm install @porsche-design-system/components-react
```

If you prefer [Yarn](https://yarnpkg.com), use the following command instead:

```bash
yarn add @porsche-design-system/components-react
```

### Usage

After adding the `@porsche-design-system/components-react` package to your project, you've to extend the standard React
setup by the `PorscheDesignSystemProvider`.

```tsx
// index.tsx

import ReactDOM from 'react-dom';
import { PorscheDesignSystemProvider } from '@porsche-design-system/components-react';
import './index.css';
import { App } from './App';

ReactDOM.render(
  <React.StrictMode>
    <PorscheDesignSystemProvider>
      <App />
    </PorscheDesignSystemProvider>
  </React.StrictMode>,
  document.getElementById('root')
);
```

Change your App file to use at least one Porsche Design System component, for example:

```tsx
// App.tsx

import { PButton } from '@porsche-design-system/components-react';

export const App = (): JSX.Element => (
  <div className="App">
    <PButton>Some label</PButton>
  </div>
);
```

## Methodology

Our goal is to provide easy-to-use and well-documented components so that developers donâ€™t need to worry about the
implementation but can focus on easily creating qualitative and consistent digital Porsche experiences. We ensure that
our components are made for everyone and meet latest quality standards of usability, accessibility, performance and
compatibility. In some points the components are built restrictive to define consistent standards for diverse Porsche
applications but ensure enough flexibility to meet different context requirements.

## License

- See **Custom License** within npm package
