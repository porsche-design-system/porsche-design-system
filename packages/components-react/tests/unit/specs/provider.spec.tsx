import { render } from '@testing-library/react';
import { PButton, PorscheDesignSystemProvider } from '../../../projects/components-wrapper/src';
import { getMergedClassName } from '../../../projects/components-wrapper/src/provider';
import { testSnapshot } from '../helpers';
import { useState } from 'react';
import userEvent from '@testing-library/user-event';
import * as pds from '@porsche-design-system/components-js';

jest.mock('@porsche-design-system/components-js', () => ({
  load: jest.fn(),
}));

describe('PorscheDesignSystemProvider', () => {
  it('should render unprefixed components', () => {
    testSnapshot(
      <PorscheDesignSystemProvider>
        <PButton>Some Button</PButton>
      </PorscheDesignSystemProvider>
    );
  });

  it('should render prefixed components', () => {
    testSnapshot(
      <PorscheDesignSystemProvider prefix="my-prefix">
        <PButton>Some Button</PButton>
      </PorscheDesignSystemProvider>
    );
  });

  it('should support changing prefix at runtime', () => {
    const Sample = (): JSX.Element => {
      const [prefix, setPrefix] = useState('my-prefix');

      return (
        <PorscheDesignSystemProvider prefix={prefix}>
          <PButton data-testid="button" onClick={() => setPrefix('new-prefix')}>
            Some Button
          </PButton>
        </PorscheDesignSystemProvider>
      );
    };

    const { container, getByTestId } = render(<Sample />);
    const button = getByTestId('button');

    expect(container).toMatchSnapshot();
    expect(pds.load).toHaveBeenCalledTimes(1);
    expect(pds.load).toHaveBeenCalledWith({ prefix: 'my-prefix' });

    userEvent.click(button);

    expect(container).toMatchSnapshot();
    expect(pds.load).toHaveBeenCalledTimes(2);
    expect(pds.load).toHaveBeenCalledWith({ prefix: 'new-prefix' });
  });

  it('should render components wrapped with nested provider and prefix', () => {
    testSnapshot(
      <PorscheDesignSystemProvider prefix="my-prefix">
        <PButton>Some Button</PButton>
        <PorscheDesignSystemProvider prefix="another-prefix">
          <PButton>Some Button</PButton>
          <PorscheDesignSystemProvider>
            <PButton>Some Button</PButton>
          </PorscheDesignSystemProvider>
        </PorscheDesignSystemProvider>
      </PorscheDesignSystemProvider>
    );
  });

  it('should throw error if not provided', () => {
    let error = '';
    const spy = jest.spyOn(global.console, 'error').mockImplementation(() => {});

    try {
      render(<PButton>Some Button</PButton>);
    } catch (e) {
      error = e.message;
    }

    expect(error).toBe('It appears the <PorscheDesignSystemProvider /> is missing. Make sure to wrap your App in it.');
    spy.mockRestore();
  });
});

describe('getMergedClassName', () => {
  test.each`
    domClasses | oldClassName   | newClassName        | expected
    ${''}      | ${undefined}   | ${''}               | ${''}
    ${''}      | ${''}          | ${undefined}        | ${''}
    ${''}      | ${undefined}   | ${undefined}        | ${''}
    ${''}      | ${''}          | ${''}               | ${''}
    ${''}      | ${''}          | ${'old1'}           | ${'old1'}
    ${''}      | ${'old1'}      | ${''}               | ${''}
    ${''}      | ${'old1'}      | ${'old1'}           | ${'old1'}
    ${''}      | ${'old1 old2'} | ${'old1'}           | ${'old1'}
    ${''}      | ${'old1 old2'} | ${'old1 old2'}      | ${'old1 old2'}
    ${''}      | ${'old1 old2'} | ${'old1 new1'}      | ${'old1 new1'}
    ${''}      | ${'old1 old2'} | ${'old1 old2 new1'} | ${'old1 old2 new1'}
    ${'dom1'}  | ${''}          | ${'new1'}           | ${'new1 dom1'}
    ${'dom1'}  | ${'old1'}      | ${'old1'}           | ${'old1 dom1'}
    ${'dom1'}  | ${'old1'}      | ${'old1 new1'}      | ${'old1 new1 dom1'}
    ${'dom1'}  | ${'old1 old2'} | ${'old1'}           | ${'old1 dom1'}
    ${'dom1'}  | ${'old1 old2'} | ${'old1 new1'}      | ${'old1 new1 dom1'}
  `(
    "should be called with ('$domClasses', '$oldClassName', '$newClassName') and return '$expected'",
    ({ domClasses, oldClassName, newClassName, expected }) => {
      const result = getMergedClassName(domClasses, oldClassName, newClassName);
      expect(result).toContain(expected);
    }
  );
});
