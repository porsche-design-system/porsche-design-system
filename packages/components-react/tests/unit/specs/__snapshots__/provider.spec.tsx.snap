// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PorscheDesignSystemProvider should render components wrapped with nested provider and prefix 1`] = `
<div>
  <my-prefix-p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
  >
    Some Button
  </my-prefix-p-button>
  <another-prefix-p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
  >
    Some Button
  </another-prefix-p-button>
  <p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
  >
    Some Button
  </p-button>
</div>
`;

exports[`PorscheDesignSystemProvider should render prefixed components 1`] = `
<div>
  <my-prefix-p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
  >
    Some Button
  </my-prefix-p-button>
</div>
`;

exports[`PorscheDesignSystemProvider should render unprefixed components 1`] = `
<div>
  <p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
  >
    Some Button
  </p-button>
</div>
`;

exports[`PorscheDesignSystemProvider should support changing prefix at runtime 1`] = `
<div>
  <my-prefix-p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
    data-testid="button"
  >
    Some Button
  </my-prefix-p-button>
</div>
`;

exports[`PorscheDesignSystemProvider should support changing prefix at runtime 2`] = `
<div>
  <new-prefix-p-button
    class="pds-skeleton--theme-light pds-skeleton--variant-secondary"
    data-testid="button"
  >
    Some Button
  </new-prefix-p-button>
</div>
`;
